### PR TITLE
[CUDNN][SDPA] Experimental cuDNN Flash Attention v2 Inference

### PR DESCRIPTION
**Pull request by eqy**
#113713

Going to clean up some of the checks and will remove draft status after.
Can be tested on SM80+ with `TORCH_CUDNN_MHA_ENABLED=1`.

CC @drisspg @ptrblck

### CODE REVIEW

File: torch/backends/cuda/__init__.py [hunk 1]
```diff
@@ -304,12 +304,30 @@ def can_use_efficient_attention(params: SDPAParams, debug: bool = False) -> bool
...
+    """
+    torch._C._set_sdp_use_cudnn(enabled)
+
 
 @contextlib.contextmanager
 def sdp_kernel(
     enable_flash: bool = True,
     enable_math: bool = True,
     enable_mem_efficient: bool = True,
+    enable_cudnn: bool = True,
```
Comment (drisspg) [inline torch/backends/cuda/__init__.py:330]: I kinda want to deprecate this function and change it to have a different api: https://github.com/pytorch/pytorch/pull/114689

But need to figure out how best to do this. 

All that being said this is the right way to do it for now

**Comment by drisspg** [comment]
Also feel free to ping me when you think I should do a review

File: aten/src/ATen/cuda/Exceptions.h [hunk 2]
```diff
@@ -21,6 +21,15 @@ class CuDNNError : public c10::Error {
 
 }  // namespace c10
 
+#define AT_CUDNN_FRONTEND_CHECK(EXPR, ...)                                                      \
```
Comment (drisspg) [inline aten/src/ATen/cuda/Exceptions.h:24]: Is this a result of frontend changing to 9.0?

File: aten/src/ATen/native/cudnn/MHA.cpp [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+}
+
+void update(const KeyType& key, T& results) {
+  engine_cache.erase(key);
+  engine_cache.emplace(key, std::move(results));
+}
+
+};
+
+// @eqy: use thread local caches as cuDNN Execution Plans are not guaranteed to be thread safe across all engines
```
Comment (drisspg) [inline aten/src/ATen/native/cudnn/MHA.cpp:147]: this comment seems at odds with the one mentioned here: https://github.com/pytorch/pytorch/pull/115663/files#diff-3fdc8929e8ab637660e7f8a73224a364bf57429e920aa4235305a2a0063b82cdR130

File: aten/src/ATen/native/cudnn/MHA.cpp [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+thread_local MHAGraphCache<graph_and_tensors, MHACacheKeyWrapper> mhagraphcache;
+
+auto build_graph_and_tensors(int64_t b,
+                                int64_t h,
+                                int64_t s_q,
+                                int64_t s_kv,
+                                int64_t d,
+                                float scaling_factor,
+                                bool return_softmaxstats,
+                                        bool is_causal,
```
Comment (drisspg) [inline aten/src/ATen/native/cudnn/MHA.cpp:158]: nit: weird formatting

File: aten/src/ATen/native/cudnn/MHA.cpp [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+                        >;
+
+#define MAX_MHA_DIM 4
+
+struct MHAParams {
+  c10::DeviceIndex device_id;
+  fe::DataType_t dataType;
+  int q_dim[MAX_MHA_DIM];
+  int k_dim[MAX_MHA_DIM];
+  int v_dim[MAX_MHA_DIM];
```
Comment (drisspg) [inline aten/src/ATen/native/cudnn/MHA.cpp:77]: std array?

File: aten/src/ATen/native/cudnn/MHA.cpp [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+        .set_intermediate_data_type(fe::DataType_t::FLOAT)
+        .set_compute_data_type(fe::DataType_t::FLOAT);
+
+    auto Q = mha_graph->tensor(fe::graph::Tensor_attributes()
+                                  .set_name("Q")
+                                  .set_dim(std::vector<int64_t>(params.q_dim, params.q_dim+MAX_MHA_DIM))
+                                  .set_stride(std::vector<int64_t>(params.q_stride, params.q_stride+MAX_MHA_DIM)));
+    auto K = mha_graph->tensor(fe::graph::Tensor_attributes()
+                                  .set_name("K")
+                                  .set_dim(std::vector<int64_t>(params.k_dim, params.k_dim+MAX_MHA_DIM))
```
Comment (drisspg) [inline aten/src/ATen/native/cudnn/MHA.cpp:184]: if std array works then could also update these construxtors

File: aten/src/ATen/native/cudnn/MHA.cpp [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+                                  .set_name("V")
+                                  .set_dim(std::vector<int64_t>(params.v_dim, params.v_dim+MAX_MHA_DIM))
+                                  .set_stride(std::vector<int64_t>(params.v_stride, params.v_stride+MAX_MHA_DIM)));
+    auto attn_scale = mha_graph->tensor(fe::graph::Tensor_attributes()
+                                       .set_name("attn_scale")
+                                       .set_dim({1, 1, 1, 1})
+                                       .set_stride({1, 1, 1, 1})
+                                       .set_is_pass_by_value(true)
+                                       .set_data_type(fe::DataType_t::FLOAT));
+    //auto bias = mha_graph->tensor(fe::graph::Tensor_attributes()
```
Comment (drisspg) [inline aten/src/ATen/native/cudnn/MHA.cpp:196]: Maybe leave a TODO: saying will work on bias in a follow up, if thats true

File: aten/src/ATen/native/cudnn/MHA.cpp [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+    //                         .set_name("bias")
+    //                         .set_dim({b, 1, s_q, s_kv})
+    //                         .set_stride({s_q * s_kv, s_q * s_kv, s_kv, 1}));
+    auto seed = mha_graph->tensor(fe::graph::Tensor_attributes()
+                                     .set_name("Seed")
```
Comment (drisspg) [inline aten/src/ATen/native/cudnn/MHA.cpp:201]: Another dumb Q: are these names meaningful? like do they need to match an api for calling the function?

File: aten/src/ATen/native/cudnn/MHA.cpp [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+                                                          .set_name("flash_attention")
+                                                          .set_is_inference(return_softmaxstats == false)
+                                                          .set_causal_mask(is_causal)
+                                                          .set_attn_scale(attn_scale)
+                                                          .set_dropout(dropout_probability, seed, offset);
+    // Optional bias in flash attention is only supported 8.9.3 onwards
+    if (cudnnGetVersion() >= 8904) {
+        //scaled_dot_product_flash_attention_options.set_alibi_mask(true);
+    }
+
```
Comment (drisspg) [inline aten/src/ATen/native/cudnn/MHA.cpp:220]: Does this only support certain bias types?

File: aten/src/ATen/native/cudnn/MHA.cpp [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+                    const Tensor& v,
+                    Tensor& softmaxstats,
+                    Tensor& o,
+                    Tensor& dropoutseed,
+                    Tensor& dropoutoffset) {
+
+    cudnnHandle_t handle = getCudnnHandle();
+    o = at::empty_strided({b, h, s_q, d}, {s_q * h * d, d, h * d, 1}, q.options());
+    if (return_softmaxstats) {
+      // TODO(eqy): fix strides
```
Comment (drisspg) [inline aten/src/ATen/native/cudnn/MHA.cpp:277]: I assume this stride fix is only needed for backward support right? so is okay for this PR?

File: aten/src/ATen/native/transformers/attention.cpp [hunk 11]
```diff
@@ -632,6 +633,14 @@ Tensor scaled_dot_product_attention(
   sdp::SDPBackend backend = static_cast<sdp::SDPBackend>(choice_int);
   c10::optional<Tensor> attn_mask = convert_boolean_attn_mask(attn_mask_, query_.dtype());
   switch (backend) {
+    case sdp::SDPBackend::cudnn: {
+      bool compute_logsumexp =
```
[the remark this answers is not in the document]
Comment (eqy) [reply aten/src/ATen/native/transformers/attention.cpp:637]: I was thinking about interoperability in this case--could we have a training case where the forward pass is done by cuDNN but the backward is handled by another SDP backend?

**Comment by drisspg** [comment]
One other random question, if I wanted to play around with this early, is there any specific build options I should be using? And any specific version of cudnn I need to have?

File: aten/src/ATen/native/transformers/attention.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (eqy) [reply aten/src/ATen/native/transformers/attention.cpp:637]: It is already supported, so a follow-up PR would add the backward kernel (which should look very similar to what was added in this PR).

**Comment by eqy** [comment]
> One other random question, if I wanted to play around with this early, is there any specific build options I should be using? And any specific version of cudnn I need to have?

I don't think you would need any specific build options---cuDNN >= 8.9 would probably work best. The only other requirement should be cuDNN frontend 1.0, but that tag was updated in the `cudnn-frontend` submodule a while ago.

File: torch/csrc/Module.cpp [hunk 12]
```diff
@@ -1238,6 +1256,11 @@ static PyMethodDef TorchMethods[] = { // NOLINT
      METH_NOARGS,
      nullptr},
     {"_set_sdp_use_math", THPModule_setSDPUseMath, METH_O, nullptr},
+    {"_get_cudnn_sdp_enabled",
```
Comment (drisspg) [inline torch/csrc/Module.cpp:1259]: There should be another bypinded struct in this file that looks like: 

``` C++
  py::enum_<sdp::SDPBackend>(
      py_module,
      "_SDPBackend",
      "Enum class for the scaled dot product attention backends\n\n... warning:: This class is in beta and subject to change.")
      .value("ERROR", sdp::SDPBackend::error)
      .value("MATH", sdp::SDPBackend::math)
      .value("FLASH_ATTENTION", sdp::SDPBackend::flash_attention)
      .value("EFFICIENT_ATTENTION", sdp::SDPBackend::efficient_attention);
      ```
      
      
      can you also update the enum here?

**Review by drisspg — COMMENTED** [review]
Left one comment, about updating the enum struct binding

I think besides the only other thing that would great is to add some more comprehensive testing. Normally we compare against the math path for both forward and backwards, I think we should analagous test for forward only comparing the math path. 

As well there is a TestSDPAFailure Modes it would be good to right some checks for the cudnn path on verify the new sdp_utils can_use_cudnn logic

File: aten/src/ATen/native/transformers/cuda/sdp_utils.cpp [hunk 13]
```diff
@@ -303,6 +303,66 @@ bool check_all_tensors_on_device(sdp_params const& params, bool debug) {
...
+  }
+  return query_layout_ok && key_layout_ok && value_layout_ok;
+}
+
+static bool check_cudnn_compute_capability(sdp_params params, bool debug) {
+  auto dprops = at::cuda::getCurrentDeviceProperties();
+  return dprops->minor == 0 && dprops->major >= 8;
+}
+
+inline bool can_use_cudnn_attention(sdp_params const& kernel_params, bool print_debug) {
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.cpp:350]: are all the check being run?

File: test/expect/HasDecompTest.test_has_decomposition.expect [hunk 14]
```diff
@@ -465,6 +465,7 @@ aten::_resize_output.out
 aten::_resize_output_
 aten::_sample_dirichlet
 aten::_sample_dirichlet.out
```
Comment (drisspg) [inline test/expect/HasDecompTest.test_has_decomposition.expect:467]: might need to be some other updates, analogous to: https://github.com/pytorch/pytorch/pull/117390

**Comment by drisspg** [comment]
Not to backseat quarterback, but would you be open to applying these changes: https://gist.github.com/drisspg/d51f74b7bb4e7986efcf9b10dc190093

Makes the sdp_utils logic more inlined with the rest of the checks as well as adds some more checks I encountered when playing around. 

That being said: I think the cudnn layout check warning could be more helpful/informative. I am using this as the driver program, and tbh I am not entirely sure what the format should be for the inputs

```Python
import torch
import torch.nn.functional as F
from torch.backends.cuda import sdp_kernel
qkv = torch.rand(8,8,64,3*64,device="cuda")
q,k,v = qkv.chunk(3,dim=-1)
with sdp_kernel(False, False, False):
    F.scaled_dot_product_attention(q,k,v, is_causal=True)
```

**Comment by eqy** [comment]
> Not to backseat quarterback, but would you be open to applying these changes: https://gist.github.com/drisspg/d51f74b7bb4e7986efcf9b10dc190093
> 
> Makes the sdp_utils logic more inlined with the rest of the checks as well as adds some more checks I encountered when playing around.
> 
> That being said: I think the cudnn layout check warning could be more helpful/informative. I am using this as the driver program, and tbh I am not entirely sure what the format should be for the inputs
> 
> ```python
> import torch
> import torch.nn.functional as F
> from torch.backends.cuda import sdp_kernel
> qkv = torch.rand(8,8,64,3*64,device="cuda")
> q,k,v = qkv.chunk(3,dim=-1)
> with sdp_kernel(False, False, False):
>     F.scaled_dot_product_attention(q,k,v, is_causal=True)
> ```

Sure, the current checks are really placeholders that need to be revisited anyway, as I haven't synchronized with the cuDNN team on expected current and future support. I was hoping to revisit the checking logic after the backwards version of this PR was done so that we could tackle expected support and testing in one fell swoop.

File: aten/src/ATen/native/transformers/cuda/sdp_utils.cpp (hunk 13 above)
Comment (eqy) [reply aten/src/ATen/native/transformers/cuda/sdp_utils.cpp:350]: These are not being run at the moment as the intention is to revisit them in tandem with backward support when that is done---currently the only real check is to have the environment variable which allows us to definitively check what is actually unsupported vs. just failing a check.

**Comment by drisspg** [comment]
Okay no worries; updated the lil exercising script:
``` Python
import torch
import torch.nn.functional as F
head_dim = 64
num_heads = 8
embed_dim = 3 * head_dim * num_heads
seq_len = 64
batch_size = 8

projection = torch.rand(batch_size, seq_len, embed_dim, device="cuda", dtype=torch.bfloat16)
q,k,v = projection.split(head_dim * num_heads, dim=-1)
q = q.view(batch_size, seq_len, num_heads, head_dim).transpose(1,2)
k = k.view(batch_size, seq_len, num_heads, head_dim).transpose(1,2)
v = v.view(batch_size, seq_len, num_heads, head_dim).transpose(1,2)

from torch.backends.cuda import sdp_kernel

with sdp_kernel(False, False, False):
    F.scaled_dot_product_attention(q,k,v, is_causal=True)
    ```
    This throws
    ```
    RuntimeError: cuDNN Frontend error: The dim for input_names::Q is invalid
    ```

**Comment by drisspg** [comment]
@eqy Okay I think that makes sense, lets just apply the changes(make sure that builds I might have forgotten a closing bracket in the patch) and then we can land this. I was going to say lets update the tracking doc but seems like you are ahead of me! But if we can get through CI/CD I think lets land to get some more momentum, thanks again for all the hard work!

**Review by drisspg — APPROVED** [review]

**Comment by PaliC** [comment]
@pytorchbot revert -m "Unfortunately, this pr breaks cuda builds internally" -c "ghfirst"
On the bright side @drisspg is working on updating cuda internally which seems to fix the problem, so you should be able to reland after his changes are in.

**Comment by eqy** [comment]
@drisspg Are we ok to retry soon?

**Comment by drisspg** [comment]
Commenting for posterity, this is still blocked by an internal upgrade required for cudnn_frontend v1.0

**Comment by izaitsevfb** [comment]
@pytorchbot merge -i

Landed internally as D53716382

**Comment by Skylion007** [comment]
We should probably update our cudnn binaries as most of the recent changes has been improving this flash attention kernal.

**Comment by yhyang201** [comment]
I would like to ask a question, and I hope it doesn't cause any inconvenience:
I noticed that MHA.cpp caches the built graph. Does this mean that if the cache hit rate is low, the performance of the cudnn backend's SDPA could be significantly degraded? (Because a lot of time would be spent on building the graph.)

**Comment by eqy** [comment]
> I would like to ask a question, and I hope it doesn't cause any inconvenience: I noticed that MHA.cpp caches the built graph. Does this mean that if the cache hit rate is low, the performance of the cudnn backend's SDPA could be significantly degraded? (Because a lot of time would be spent on building the graph.)

yes, although mainly because the graph rebuild would incur JIT compilation in addition to some CPU overhead

**Comment by yhyang201** [comment]
> > I would like to ask a question, and I hope it doesn't cause any inconvenience: I noticed that MHA.cpp caches the built graph. Does this mean that if the cache hit rate is low, the performance of the cudnn backend's SDPA could be significantly degraded? (Because a lot of time would be spent on building the graph.)
> 
> yes, although mainly because the graph rebuild would incur JIT compilation in addition to some CPU overhead

Thank you very much for your prompt response!

May I ask the reason why cache is not used in `run_cudnn_SDP_fprop_nestedtensor`?

**Pull request closed without merging**